crates/autopilot/src/risk_adjusted_rewards.rs

```diff
@@ -181,9 +203,85 @@ fn uncapped_reward_eth_atoms(
     (profit + cost) / (1. - revert_probability) - cost
 }
 
+/// A caching wrapper over [NativePriceEstimating] used to estimate and cache
+/// CoW token price. Implemented to enable the [Calculator] produce rewards even
+/// when a fresh estimate is not available.
+struct BestEffortCowPriceEstimator {
+    inner: Arc<dyn NativePriceEstimating>,
+    cow_token: H160,
+    /// Cached value and a timestamp.
+    fallback_cache: Mutex<Option<(f64, Instant)>>,
+    /// How long is the value valid after being set.
+    max_age: Duration,
+}
+
+impl BestEffortCowPriceEstimator {
+    fn new(inner: Arc<dyn NativePriceEstimating>, cow_token: H160, max_age: Duration) -> Self {
+        Self {
+            inner,
+            cow_token,
+            fallback_cache: Mutex::new(None),
+            max_age,
+        }
+    }
+
+    /// Attempts to use the inner estimator to get a fresh price estimate.
+    /// If that fails, attempts to return the value from cache.
+    async fn get_price(&self, current_time: Instant) -> NativePriceEstimateResult {
+        let fresh = self
+            .inner
+            .estimate_native_prices(std::slice::from_ref(&self.cow_token))
+            .next()
+            .await
+            .unwrap()
+            .1;
+
+        match fresh {
+            Ok(price) => {
+                self.set_cached(price, current_time);
+                Ok(price)
+            }
+            Err(error) => {
+                let price = self.get_cached(current_time).ok_or(error);
+
+                if price.is_ok() {
+                    tracing::warn!("Using an old CoW token price from a fallback cache");
```

Comment:
Can we include the original error in the log message? Might be useful for debugging.

Reply:
Sure :+1: